gems/sorbet-runtime/lib/types/private/class_utils.rb

```diff
@@ -98,6 +98,9 @@ def self.replace_method(mod, name, &blk)
     T::Configuration.without_ruby_warnings do
       T::Private::DeclState.current.without_on_method_added do
         mod.send(:define_method, name, &blk) # rubocop:disable PrisonGuard/UsePublicSend
+        if blk.arity < 0 && mod.respond_to?(:ruby2_keywords, true)
+          mod.send(:ruby2_keywords, name)
+        end
```

Comment:
There are a couple dozen calls to `define_method` in `sorbet-runtime`. Why are these the only two that need `ruby2_keywords`?

Reply:
Because that's the only two that sometimes do splat (`*args`) delegation.  

That's also why I check for negative arity.

Reply:
From a quick grep, it looks like these might be important to use ruby2_keywords with too:

https://github.com/sorbet/sorbet/blob/bbfacab5b1d253ab91a82a3efa17380e5a479cc4/gems/sorbet-runtime/lib/types/private/abstract/declare.rb#L37-L42

https://github.com/sorbet/sorbet/blob/bbfacab5b1d253ab91a82a3efa17380e5a479cc4/gems/sorbet-runtime/lib/types/props/has_lazily_specialized_methods.rb#L69-L73

Maybe this one, but i'm not sure if redefining an existing method preserves the `ruby2_keywords` setting of that method:

https://github.com/sorbet/sorbet/blob/bbfacab5b1d253ab91a82a3efa17380e5a479cc4/gems/sorbet-runtime/lib/types/private/class_utils.rb#L36

I'd appreciate if you did a look of your own and made a call on each of them

Reply:
Ok, so:

  -  `abstract/declare.rb`: I was able to generate a test case that would trigger the warnings, it's very contrived but better safe than sorry.
  - `props/has_lazily_specialized_methods.rb` that appear to be needed to. But I'm unsure how to exercise it in the test suite.
  - `private/class_utils.rb`: I don't think it's required because `define_method` is called with a `Method` instance, so the splat arguments handling should be preserved.

Reply:
Awesome, thanks for looking into that!

Reply:
@djudd-stripe Do you have a second to chime on on the `has_lazily_specialized_methods` case?

The context is:

- in Ruby 2.7, keyword args are handled differently, and there will be warnings if a hash is used in place of keyword args
- casperisfine has worked around this in a way that should silence the warnings, but we don't know how to write a test that would verify that

Can you think of a test case that would trigger the `*args` here to raise a warning because it's trying to forward a hash instead of keyword arguments directly?

https://github.com/sorbet/sorbet/blob/bbfacab5b1d253ab91a82a3efa17380e5a479cc4/gems/sorbet-runtime/lib/types/props/has_lazily_specialized_methods.rb#L71

Reply:
We only use this for `serialize` and `deserialize` right now and they don't use keyword arguments, so to test it I think you'd have to construct a special-purpose test T::Props::Plugin using `Class.new` or something. I'm not sure it's worth the effort since this is not really intended to be a public extension point; maybe we should try to ban use of keyword arguments here instead?